extensions/amp-auto-lightbox/0.1/carousel-criteria.js

```diff
@@ -0,0 +1,294 @@
+  'I': true,
+  'IMG': true,
+  'LABEL': true,
+  'LEGEND': true,
+  'NOBR': true,
+  'NOSCRIPT': true,
+  'O:P': true,
+  'P': true,
+  'PRE': true,
+  'Q': true,
+  'RP': true,
+  'RT': true,
+  'RTC': true,
+  'RUBY': true,
+  'S': true,
+  'SCRIPT': true,
+  'SECTION': true,
+  'SMALL': true,
+  'SPAN': true,
+  'STRIKE': true,
+  'STRONG': true,
+  'SUB': true,
+  'SUMMARY': true,
+  'SUP': true,
+  'TIME': true,
+  'U': true,
+  'VAR': true,
+  'WBR': true,
+
+  // The following could potentially be admissible but are either unlikely
+  // to be used where auto-lightbox is desired, or are obsolete.
+  // This list is kept but commented-out to keep track of possible elements to
+  // consider.
+  // 'ADDRESS': true,
+  // 'ARTICLE': true,
+  // 'ASIDE': true,
+  // 'CODE': true,
+  // 'DD': true,
+  // 'DEL': true,
+  // 'DFN': true,
+  // 'DL': true,
+  // 'DT': true,
+  // 'FOOTER': true,
+  // 'HEADER': true,
+  // 'INS': true,
+  // 'KBD': true,
+  // 'LISTING': true,
+  // 'MAIN': true,
+  // 'MARK': true,
+  // 'METER': true,
+  // 'MULTICOL': true,
+  // 'NAV': true,
+  // 'NEXTID': true,
+  // 'SAMP': true,
+  // 'SPACER': true,
+};
+
+export class CarouselCriteria {
+  /**
+   * @param {!Element} element
+   * @return {!Promise<boolean>}
+   */
+  static meetsAll(element) {
+    const win = toWin(element.ownerDocument.defaultView);
+
+    if (!isExperimentOn(win, 'amp-auto-lightbox-carousel')) {
+      return resolveFalse();
+    }
+
+    const slides = element.querySelectorAll('.amp-carousel-slide');
+    const images = element.querySelectorAll('amp-img');
+
+    if (images.length < 1) {
+      return resolveFalse();
+    }
+
+    if (slides.length != images.length) {
+      return resolveFalse();
+    }
+
+    let promise = resolveTrue();
+
+    iterateCursor(slides, slide => {
+      promise = promise.then(previousWasAccepted => {
+        if (!previousWasAccepted) {
+          return false;
+        }
+        return SlideCriteria.meetsAll(slide);
+      });
+    });
+
+    return promise;
+  }
+}
+
+class SlideCriteria {
+
+  /**
+   * @param {!Element} element
+   * @return {!Promise<boolean>}
+   */
+  static meetsAll(element) {
+    if (element.tagName == 'AMP-IMG') {
+      // Already traversed up the tree, we only need to check the immediate
+      // action.
+      const actionLookupStopAt = dev().assertElement(element.parentElement);
+      const actions = Services.actionServiceForDoc(element);
+      return tryResolve(() =>
+        !actions.hasResolvableAction(element, 'tap', actionLookupStopAt));
+    }
+
+    const img = element.querySelector('amp-img');
+    if (!img) {
+      return resolveFalse();
+    }
+
+    const slideMeetsSizingPromise =
+        SlideCriteria.meetsSizingCriteria(img, element);
+
+    return slideMeetsSizingPromise.then(slideMeetsSizingCriteria => {
+      if (!slideMeetsSizingCriteria) {
+        return false;
+      }
+      return SlideCriteria.meetsAllSync(element, /* depth */ 0);
```

Comment:
small suggestion: Using a helper function or default parameter would allow callers of `meetsAllSync` to omit passing `0` for the depth.

Reply:
Or maybe I don't understand the purpose of `meetsAllSync` vs `meetsAll`